### PR TITLE
Distributed Gradient Boosted Trees Regression

### PR DESCRIPTION
**Pull request by Alexsandruss**
## Pull request: Distributed Gradient Boosted Trees Regression
#  Description
Implementation of Distributed Gradient Boosted Trees Regression with inexact split mode for Intel® DAAL.

Pull request consist of:

 - Algorithm implementation
 - C++ example and sample
 - Java interfaces and example

### CODE REVIEW

**Comment by ShvetsKS** [comment]
make test

**Comment by ShvetsKS** [comment]
please fix all issue related to Precommit:
http://intel-ci.intel.com/e9fc6bd6-e569-f1a8-bc61-8cdcd4b723a1

File: algorithms/kernel/dtrees/dtrees_feature_type_helper.i [hunk 1]
```diff
@@ -340,20 +356,99 @@ services::Status IndexedFeatures::init(const NumericTable& nt, const FeatureType
...
+    {
+        //in case of single thread no need to allocate
+        TlsTask* task = tlsData.local();
+        DAAL_CHECK_THR(task, services::ErrorMemoryAllocationFailed);
+        safeStat |= task->makeIndex(const_cast<NumericTable&>(nt), _entries[iCol], _data + iCol*nRows(), iCol, nRows(),
+            featureTypes->isUnordered(iCol), _minFeatureValues + iCol, true);
+        if (pBimPrm)
+        {
+            const size_t maxBins = pBimPrm->maxBins;
+            size_t *localBins = binsPtr[iCol]->get();
```
Comment (averbukh) [inline algorithms/kernel/dtrees/dtrees_feature_type_helper.i]: size_t * const ?

File: algorithms/kernel/dtrees/dtrees_feature_type_helper.i [hunk 2]
```diff
@@ -340,20 +356,99 @@ services::Status IndexedFeatures::init(const NumericTable& nt, const FeatureType
...
+            for (size_t iBin = 0; iBin < maxBins; ++iBin)
+                localBins[iBin] = (task->getBins())[iBin];
+        }
+    });
+    tlsData.reduce([&](TlsTask* task)-> void
+    {
+        if(_maxNumIndices < task->maxNumDiffValues)
+            _maxNumIndices = task->maxNumDiffValues;
+        delete task;
```
Comment (averbukh) [inline algorithms/kernel/dtrees/dtrees_feature_type_helper.i]: SDL violation

File: algorithms/kernel/dtrees/dtrees_model.cpp [hunk 3]
```diff
@@ -163,9 +163,8 @@ services::Status createTreeInternal(data_management::DataCollectionPtr& serializ
     const size_t nTrees = (*(serializationData)).size();
     while(isNotEmptyTreeTable && (treeId < nTrees))
     {
+        treeId++;
```
Comment (averbukh) [inline algorithms/kernel/dtrees/dtrees_model.cpp]: Use prefix increment instead

File: algorithms/kernel/dtrees/gbt/gbt_train_partition.i [hunk 4]
```diff
@@ -133,6 +133,112 @@ public:
     {
     }
 
+    static DAAL_INT doPartitionIdxWithStride(IndexType n, RowIndexType* aIdx, const BinIndexType* indexedFeature, bool featureUnordered,
+        RowIndexType idxFeatureValueBestSplit, RowIndexType* buffer, RowIndexType &nLeft, size_t nFeatures = 1, size_t featureIndex = 0,
+        bool isDistributed = false)
+    {
+        DAAL_INT iRowSplitVal = -1;
+
+        const size_t maxNBlocks = 56;
```
Comment (averbukh) [inline algorithms/kernel/dtrees/gbt/gbt_train_partition.i]: Add comment

File: algorithms/kernel/dtrees/gbt/gbt_train_partition.i [hunk 5]
```diff
@@ -133,6 +133,112 @@ public:
...
+        size_t sizeOfBlock = 2048;
+        size_t nBlocks = n/sizeOfBlock;
+        nBlocks += !!(n - nBlocks*sizeOfBlock);
+
+        if (nBlocks > maxNBlocks)
+        {
+            nBlocks = maxNBlocks;
+            sizeOfBlock = n/nBlocks + !!(n%nBlocks);
```
Comment (averbukh) [inline algorithms/kernel/dtrees/gbt/gbt_train_partition.i]: What about alignment of the block? Add comment.

File: algorithms/kernel/dtrees/gbt/gbt_train_partition.i [hunk 6]
```diff
@@ -133,6 +133,112 @@ public:
...
+        }
+
+        RowIndexType part_high_left[maxNBlocks];
```
Comment (averbukh) [inline algorithms/kernel/dtrees/gbt/gbt_train_partition.i]: maxNBlocks should be constexpr

File: algorithms/kernel/dtrees/gbt/gbt_train_partition.i [hunk 7]
```diff
@@ -133,6 +133,112 @@ public:
...
+        RowIndexType part_high_right[maxNBlocks];
```
Comment (averbukh) [inline algorithms/kernel/dtrees/gbt/gbt_train_partition.i]: Should be initialized to follow SDL?

File: algorithms/kernel/dtrees/gbt/gbt_train_partition.i [hunk 8]
```diff
@@ -133,6 +133,112 @@ public:
...
+                }
+            }
+
+            part_high_left[iBlock] = iLeft;
+            part_high_right[iBlock] = iRight;
+        });
+
+        nLeft = 0;
+
+        for (size_t iBlock = 0; iBlock < nBlocks; iBlock++)
```
Comment (averbukh) [inline algorithms/kernel/dtrees/gbt/gbt_train_partition.i]: Fix coding style

File: algorithms/kernel/dtrees/gbt/gbt_train_partition.i [hunk 9]
```diff
@@ -133,6 +133,112 @@ public:
...
+        });
+
+        if (isDistributed)
+        {
+            iRowSplitVal = 0;
+        }
+        else
+        {
+            RowIndexType i = 0;
+            while(indexedFeature[aIdx[i] * nFeatures + featureIndex] != idxFeatureValueBestSplit) i++;
```
Comment (averbukh) [inline algorithms/kernel/dtrees/gbt/gbt_train_partition.i]: Fix coding style

File: algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h [hunk 10]
```diff
@@ -0,0 +1,175 @@
+/* file: gbt_regression_init_container.h */
+/*******************************************************************************
+* Copyright 2014-2019 Intel Corporation
```
Comment (averbukh) [inline algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h]: Why 2014?

File: algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+    const NumericTablePtr ntLocalData = input->get(step1LocalData);
+    const NumericTablePtr ntLocalDependentVariables = input->get(step1LocalDependentVariables);
+
+    const HomogenNumericTable<algorithmFPType> *ntMeanDependentVariable = dynamic_cast<HomogenNumericTable<algorithmFPType>*>((partialResult->get(step1MeanDependentVariable)).get());
+    const HomogenNumericTable<size_t> *ntNumberOfRows = dynamic_cast<HomogenNumericTable<size_t>*>((partialResult->get(step1NumberOfRows)).get());
+    const HomogenNumericTable<algorithmFPType> *ntBinBorders = dynamic_cast<HomogenNumericTable<algorithmFPType>*>((partialResult->get(step1BinBorders)).get());
+    const HomogenNumericTable<size_t> *ntBinSizes = dynamic_cast<HomogenNumericTable<size_t>*>((partialResult->get(step1BinSizes)).get());
+
+    Parameter *par = static_cast<Parameter *>(_par);
```
Comment (averbukh) [inline algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h]: Should be '* const' where applicable

File: algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+DistributedContainer<step2Master, algorithmFPType, method, cpu>::~DistributedContainer()
+{
+    __DAAL_DEINITIALIZE_KERNELS();
+}
+
+template <typename algorithmFPType, Method method, CpuType cpu>
+services::Status DistributedContainer<step2Master, algorithmFPType, method, cpu>::compute()
+{
+    DistributedInput<step2Master> *input = static_cast<DistributedInput<step2Master> *>(_in);
+    DistributedPartialResultStep2 *partialResult = static_cast<DistributedPartialResultStep2 *>(_pres);
```
Comment (averbukh) [inline algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h]: Should be '* const' where applicable

File: algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+    const DataCollectionPtr dcBinBorders = input->get(step2BinBorders);
+    const DataCollectionPtr dcBinSizes = input->get(step2BinSizes);
+    const DataCollectionPtr dcMeanDependentVariable = input->get(step2MeanDependentVariable);
+    const DataCollectionPtr dcNumberOfRows = input->get(step2NumberOfRows);
+
+    size_t nNodes = dcBinBorders->size();
```
Comment (averbukh) [inline algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h]: const ?

File: algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+
+    HomogenNumericTable<algorithmFPType> *ntInitialResponse = dynamic_cast<HomogenNumericTable<algorithmFPType>*>((partialResult->get(step2InitialResponse)).get());
+    const HomogenNumericTable<algorithmFPType> *ntMergedBinBorders = dynamic_cast<HomogenNumericTable<algorithmFPType>*>((partialResult->get(step2MergedBinBorders)).get());
+    const HomogenNumericTable<size_t> *ntBinQuantities = dynamic_cast<HomogenNumericTable<size_t>*>((partialResult->get(step2BinQuantities)).get());
```
Comment (averbukh) [inline algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h]: Should be '* const' where applicable

File: algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+    const DataCollectionPtr dcBinValues = partialResult->get(step2BinValues);
+
+    Parameter *par = static_cast<Parameter *>(_par);
```
Comment (averbukh) [inline algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h]: Should be '* const' where applicable

File: algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+
+template <typename algorithmFPType, Method method, CpuType cpu>
+services::Status DistributedContainer<step3Local, algorithmFPType, method, cpu>::compute()
+{
+    DistributedInput<step3Local> *input = static_cast<DistributedInput<step3Local> *>(_in);
+    const DistributedPartialResultStep3 *partialResult = static_cast<DistributedPartialResultStep3 *>(_pres);
+    Parameter *par = static_cast<Parameter *>(_par);
+
+    const HomogenNumericTable<algorithmFPType> *ntMergedBinBorders = dynamic_cast<HomogenNumericTable<algorithmFPType>*>(input->get(step3MergedBinBorders).get());
+    const HomogenNumericTable<size_t> *ntBinQuantities = static_cast<HomogenNumericTable<size_t>*>((input->get(step3BinQuantities)).get());
```
Comment (averbukh) [inline algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h]: Should be '* const' where applicable

File: algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+    const NumericTablePtr ntLocalData = input->get(step3LocalData);
+    const HomogenNumericTable<algorithmFPType> *ntInitialResponse = dynamic_cast<HomogenNumericTable<algorithmFPType>*>(input->get(step3InitialResponse).get());
+
+    daal::services::Environment::env &env = *_env;
```
Comment (averbukh) [inline algorithms/kernel/dtrees/gbt/regression/gbt_regression_init_container.h]: Should be '* const' where applicable

**Pull request closed without merging**